### PR TITLE
Fix Flaky unittest, allocate enough space for all needed opCodes

### LINKED ISSUE #507: Code generation introduces non-deterministic failure in ct/evm_test.go
The test has experienced intermittent failures in CI when compiling unrelated changes.
Initial investigation shows that the code generator attempts to introduce 2 op codes in a code segment of length 1, this
is a scenario that the algorithm should be able to solve. 

example of failure with unrelated code changes: https://github.com/Fantom-foundation/Tosca/actions/runs/9413810813/job/25931362205?pr=506

### PR DESCRIPTION
**Pull request by LuisPH3**
What does this solve?  
The unit test would have failed because of creating a code buffer of size 1 to store 2 instructions.

What does this not solve: 
It is still not possible to additionally set the Pc to a constant direction, this would require to generate the Pc at a different point in time and account for it. This is not needed right now.

fixes #507

### CODE REVIEW

**Review by facuMH — DISMISSED** [review]
thanks for making the test more stable. nice catch on edge case

File: go/ct/gen/code.go [hunk 1]
```diff
@@ -114,6 +115,22 @@ func (g *CodeGenerator) Generate(assignment Assignment, rnd *rand.Rand) (*st.Cod
...
+	for _, constraint := range varOps {
+		opCount[constraint.op] = true
+	}
+	for _, constraint := range ops {
+		opCount[constraint.op] = true
+	}
+	minSize = max(minSize, len(opCount))
+
+	// If there are any variables that need to be bound to code, there must be at
+	// least on instruction in the resulting code.
```
Comment (HerbertJordan) [inline go/ct/gen/code.go:129]:
```suggestion
	// least one instruction in the resulting code.
```

File: go/ct/gen/code.go [hunk 2]
```diff
@@ -114,6 +115,22 @@ func (g *CodeGenerator) Generate(assignment Assignment, rnd *rand.Rand) (*st.Cod
...
 
+	// Make enough space to host all the different opCodes used in condition.
+	opCount := make(map[OpCode]bool)
+	for _, constraint := range varOps {
+		opCount[constraint.op] = true
+	}
+	for _, constraint := range ops {
+		opCount[constraint.op] = true
+	}
+	minSize = max(minSize, len(opCount))
```
Comment (HerbertJordan) [inline go/ct/gen/code.go:126]: Please create a test case in `code_test.go` that fails without this modification and passes with this change.

File: go/ct/gen/code.go [hunk 3]
```diff
@@ -114,6 +115,22 @@ func (g *CodeGenerator) Generate(assignment Assignment, rnd *rand.Rand) (*st.Cod
...
+
+	// If there are any variables that need to be bound to code, there must be at
+	// least on instruction in the resulting code.
+	if minSize == 0 && len(g.varIsCodeConstraints) > 0 {
+		minSize = 1
+	}
```
Comment (HerbertJordan) [inline go/ct/gen/code.go:132]: Please create a test case in `code_test.go` that fails without this modification and passes with this change.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: go/ct/gen/code.go (hunk 2 above)
Comment (LuisPH3) [reply go/ct/gen/code.go:126]: It is going to be pretty difficult to make deterministic, after all, the code size is a function of rand.

Comment (LuisPH3) [reply go/ct/gen/code.go:126]: image
I need some ideas about how to write a test that can call this function and deterministically test the corner cases.
I can only think about the injection of a random generator object mock, which I could influence.

Comment (HerbertJordan) [reply go/ct/gen/code.go:126]: It might help adding a upper bound for the code size as an additional constraint to the generator, narrowing down on the `st.MaxCodeSize`. This way, you can force the `size` to be 0, 1, or 2 in tests. Wdyt?

Comment (HerbertJordan) [reply go/ct/gen/code.go:126]: You might even have a fixed code-size constraint added just for testing. Something like
```
codeSize *int
```
as a simple field, that if set is used, and if `nil` is ignored.

File: go/ct/gen/code_test.go [hunk 4]
```diff
@@ -354,6 +354,142 @@ func TestCodeGenerator_ClonesCanBeUsedToResetGenerator(t *testing.T) {
 	}
 }
 
+func TestCodeGenerator_CodeIsLargeEnoughForAllConditionOps(t *testing.T) {
+
+	type constOp struct {
+		p  int
```
Comment (HerbertJordan) [inline go/ct/gen/code_test.go:360]: Please do not use abbreviation for identifiers. It makes reading the code unnecessarily hard.

File: go/ct/gen/code_test.go [hunk 5]
```diff
@@ -354,6 +354,142 @@ func TestCodeGenerator_ClonesCanBeUsedToResetGenerator(t *testing.T) {
...
+		op OpCode
+	}
+
+	type varOp struct {
+		v  string
+		op OpCode
+	}
+
+	tests := map[string]struct {
+		offset       int
```
Comment (HerbertJordan) [inline go/ct/gen/code_test.go:370]: The name `offset` does not convey any information about the purpose of this field.

Looking at the implementation below, it looks like it is the forced code length. So why not name it `codeLength` or `codeSize`?

File: go/ct/gen/code.go [hunk 6]
```diff
@@ -114,12 +118,33 @@ func (g *CodeGenerator) Generate(assignment Assignment, rnd *rand.Rand) (*st.Cod
...
+
+	// If there are any variables that need to be bound to code, there must be at
+	// least one instruction in the resulting code.
+	if minSize == 0 && len(g.varIsCodeConstraints) > 0 {
+		minSize = 1
+	}
+
+	var size int
+	if g.codeSize != nil {
+		size = *g.codeSize + minSize
```
Comment (HerbertJordan) [inline go/ct/gen/code.go:139]: This is surprising ... the field is named `codeSize` but it does not seem to put a constraint on the code size. Instead, it adds some extra "boost" to the minimum code size?

I would have thought to make it simply a strict constraint on the code size. If `codeSize` is set and it has a value, that is the size of the produced code. If the value is smaller than the `minSize`, the generator produces an `Unsatisfiable` error.

This way, the parameter has a simpler interpretation. Wdyt?

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: go/ct/gen/code.go (hunk 6 above)
Comment (LuisPH3) [reply go/ct/gen/code.go:139]: Right, that would work if the code size would be generated using a range. but instead the generated size is using this ad-hoc code:
image
I am not sure what takes precedence here, the feature of returning small code sizes seems reasonable to me. 
how would you proceed, would you add an exponential generator to the range generator tool?

**Review by HerbertJordan — APPROVED** [review]

**Review by facuMH — APPROVED** [review]

**Review by simonlechner — APPROVED** [review]
